### ISSUE TITLE
Tab autocomplete in terminal runs code?

### ISSUE THREAD

**Issue by jeanmn** [ipython/ipython#9094]
Ipython executes parts of my code when using the autocomplete functionality.

#### System info:
```bash
>> python -c "import IPython; print(IPython.sys_info())"
{'commit_hash': u'8b97549',
 'commit_source': 'repository',
 'default_encoding': 'UTF-8',
 'ipython_path': '/home/ipython/IPython',
 'ipython_version': '4.1.0-dev',
 'os_name': 'posix',
 'platform': 'Linux-3.16.0-53-generic-x86_64-with-Ubuntu-14.04-trusty',
 'sys_executable': '/home/ipython/virtual_environment/bin/python',
 'sys_platform': 'linux2',
 'sys_version': '2.7.6 (default, Jun 22 2015, 17:58:13) \n[GCC 4.8.2]'}
```

#### Example:
```ipython
In [1]: class Hello(object):                                                                                                                                            
    @property                                                                   
    def hello(self):                                                                
        print "It's me" 

In [2]: h=Hello()
```
Trying to autocomplete `h.b`, `h.hello` or `h.b(` works as I would expect. However `h.h(` or `h.hello(` results in 
```python
In [3]: h.h(It's me
Display all ... possibilities? (y or n)
....
In [3]: h.h(
```
which is resulted in an unexpected printout.

**Comment by takluyver** [comment]
We make a tradeoff here. On one hand, completing shouldn't cause arbitrary code to execute. But because Python is so flexible, almost anything can execute code, and it's hard to tell what will (and whether it will matter).

Attribute access (`a.foo`) doesn't normally cause arbitrary code to run, and when it does, it's generally expected to be something without side effects. So we allow completion to do it in order to get the object you're completing on. It won't do function calls (`a('foo')`) or item access (`a['foo']`) by default.

**Comment by jeanmn** [comment]
@takluyver Thank your for the quick response. Does this explain the `h.h(<TAB>` case?
 The class has no attribute `h`.

**Comment by takluyver** [comment]
Hmm, yeah, that's a bit weird.

**Comment by Carreau** [comment]
#10182 should be a step toward that, if provides completion using jedi that should _not_ evaluate anything

**Comment by bilderbuchi** [comment]
@Carreau using IPython 6.1.0, which contains the changes from #10182 afaict, still shows this behaviour.
In fact it got worse, all the completions listed by @jeanmn trigger either one or two printouts of "It's me", which shows that the property gets executed at least one time when completing. 
By the way, you [hint](https://github.com/pandas-dev/pandas/issues/16409#issuecomment-302966349) in  pandas-dev/pandas#16409 that that bug (the same like this one?) is a Jedi bug - do you know which one? 
cc @davidhalter

**Comment by bilderbuchi** [comment]
Ah, further digging revealed the newer #10720 and davidhalter/jedi#925, which could be behind this.

**Comment by davidhalter** [comment]
I agree, I should really get that fixed. I'm still not sure how easy it is, though. Doesn't Python's default shell autocompletion also trigger this?

**Comment by bilderbuchi** [comment]
> Doesn't Python's default shell autocompletion also trigger this?
I just tried: No, you only get the correct autocompletion to `h.hello`.

**Comment by mdeff** [comment]
Looks like this has been fixed in jedi 0.11 (davidhalter/jedi#925). At least upgrading jedi fixed the issue for me.

**Comment by dannygoldstein** [comment]
I am still seeing this issue with Jedi 0.15.1

**Comment by natezb** [comment]
Unfortunately, this issue was reintroduced with Jedi 0.15 (see https://github.com/davidhalter/jedi/issues/1063#issuecomment-520254081).

**Comment by davidhalter** [comment]
This was intentionally reintroduced, since this is something people rely on (unfortunately):

https://github.com/davidhalter/jedi/issues/1299#issuecomment-520253660

I decided to go with the "unsafe" version, because way more people complained about not having completions. I also kind of agree with this now, because a property should never be something that takes a lot of time to execute, otherwise it should probably be part of a function. It's essentially a sign of bad code.

If you don't use Jedi autocompletions will by the way also execute properties.

However I know that sometimes people rely on external code, so they cannot change it. There is still the option to set the private `jedi.Interpreter._allow_descriptor_getattr_default = False`. This will break at some point, but for people looking for a temporary workaround, this should be good enough.

IMO this issue should be closed, because the majority seems to like property execution and IMO having a public switch is not really worth it. If IPython maintainers really want to have a switch, we can probably make the boolean above public (as a param to `Interpreter`), but interest for this has been pretty low.

**Comment by natezb** [comment]
While I see the argument about properties vs functions, I can't say I fully agree.

But regardless, after testing in a plain Python REPL, I do see that it has the same behavior, so I can respect the decision. Can you briefly explain (or direct me to an explanation of) why it is necessary to run `__get__()` to complete the name of a property? It just seems odd that running `myobject.<TAB>` will execute the `__get__()` of every property and descriptor of that object. I understand it's obviously necessary if you want to complete the attributes *of* said property, but not its name. Can a property raise `AttributeError` to avoid showing up or something?

**Comment by davidhalter** [comment]
AFAIK this happens because IPython wants to know more than the name about a completion. It's probably the type that they want to know (i.e. `class` or `function`).

**Comment by bilderbuchi** [comment]
Thank you, but what about plain Python? I also just reproduced this in a fresh minimal conda environment with only Python 3.9, and see that it is in fact executing on autocompletion of `h.<TAB>`, but there is no IPython nor Jedi around:
```
(ipython-9094) bilderbuchi@osprey:~$ python
Python 3.9.1 (default, Dec 11 2020, 14:32:07) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class Hello:
...     @property
...     def hello(self):
...         print('In an executing property')
... 
>>> h = Hello()
>>> h.In an executing property
hello
In an executing property
>>> 
[4]+  Stopped                 python
(ipython-9094) bilderbuchi@osprey:~$ conda list
# packages in environment at /home/bilderbuchi/anaconda3/envs/ipython-9094:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
ca-certificates           2020.12.8            h06a4308_0  
certifi                   2020.12.5        py39h06a4308_0  
ld_impl_linux-64          2.33.1               h53a641e_7  
libedit                   3.1.20191231         h14c3975_1  
libffi                    3.3                  he6710b0_2  
libgcc-ng                 9.1.0                hdf63c60_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
ncurses                   6.2                  he6710b0_1  
openssl                   1.1.1i               h27cfd23_0  
pip                       20.3.3           py39h06a4308_0  
python                    3.9.1                hdb3f193_2  
readline                  8.0                  h7b6447c_0  
setuptools                51.0.0           py39h06a4308_2  
sqlite                    3.33.0               h62c20be_0  
tk                        8.6.10               hbc83047_0  
tzdata                    2020d                h14c3975_0  
wheel                     0.36.2             pyhd3eb1b0_0  
xz                        5.2.5                h7b6447c_0  
zlib                      1.2.11               h7b6447c_3  
```
I searched for an issue in the Python bug tracker, but could not find anything.

Also, please allow me one remark on this:
> a property should never be something that takes a lot of time to execute, otherwise it should probably be part of a function. It's essentially a sign of bad code.

At least for our part (and probably also @natezb), this is not primarily about speed, but undesirable communication/code execution (in this case communication with hardware over the wire). I sometimes use a sourcemeter instrument with ~50 properties defined, I'll have to check some time what exactly happens if I try to autocomplete on that one. Yes, that could well become a speed issue on a serial connection, but I am more concerned about accidental side effects of code that was never intentionally run by the user.

**Comment by davidhalter** [comment]
> Properties are used to implement a simple (for the user) instrument interaction API, e.g. both setting and getting e.g. an instrument.voltage, while hiding away all the communication, value validation, and other details.

I think if the voltage value is not in memory or takes time to calculate, it should be implemented with functions (get_voltage, set_voltage). This is more explicit than a property and a user will be able to understand the difference.
